### PR TITLE
initial pre-compute transitive deps issue #8

### PR DESCRIPTION
**Pull request by DevinR528**
I've started mapping out how to go about implementing this. <s/>The main/lib thing makes reusing tally.rs's structs difficult so I plan on adding transitive computation in the index crate?</s>
I can move the transitive calculating functionality into the index crate (`Resolve`, `ResolvedCrate` plus some of `Universe` functionality)?  Or maybe split cargo_tally lib into its own crate?

### CODE REVIEW

File: index/src/transitive.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use cargo_tally::Crate;
+
+#[derive(Serialize, Deserialize)]
+pub struct TranitiveCrateDeps {
+    pub name: String,
+    /// Crates that depend on this crate
+    depended_on: Vec<Someting>,
```
Comment (DevinR528) [inline index/src/transitive.rs]: Do we want to use a `Vec<Something>` or just total each crates transitive depended on count.

File: index/src/main.rs [hunk 2]
```diff
@@ -39,14 +42,19 @@ fn main() {
     }
 }
 
+// TODO ask about try_main
```
Comment (DevinR528) [inline index/src/main.rs]: do you want me to make a pr updating `main` and removing `try_main`?

File: index/src/main.rs [hunk 3]
```diff
@@ -39,14 +42,19 @@ fn main() {
...
     let crates = parse_index(&opts.index)?;
     let pb = setup_progress_bar(crates.len());
     let timestamps = compute_timestamps(repo, &pb)?;
     let crates = consolidate_crates(crates, timestamps);
-    write_json(crates)?;
+    let transitive = compute_transitive(&crates);
+
+    write_json(cargo_tally::JSONFILE, crates)?;
+    // or make a new function
+    write_json(cargo_tally::COMPFILE, transitive)?;
```
Comment (DevinR528) [inline index/src/main.rs]: Change `write_json` to accept `T: serde::Serialize` or make new `write_trans_json`

**Comment by DevinR528** [comment]
I'll start ripping out everything that can be removed from tally. Any suggestions on how to test the thing, or just run and `Ctr-c` after printing a few?

**Comment by DevinR528** [comment]
broken-serde-1
I've made some progress but I've hit a major bump. The graph I get after pre-computing transitive dependencies only accounts for events in time that represent version bumps. 

It seems to enable the `--exclude` feature and ability to compare arbitrary crates I need to keep a lot more information in the serialized struct. Am I correct in thinking that each crate needs to know every crate that depends on it at every point in time? This allows comparison without timestamp problems hopefully fixing my broken graph problem.

My next attempt will be keeping indexes into the `Vec<Crate>` so the serialized struct would have a `HashMap<timestamp, Vec<index>>` would work. 
I'm still getting comfortable with this code base its complicated but thank you for the opportunity.

File: index/src/main.rs (hunk 2 above)
Comment (dtolnay) [reply index/src/main.rs]: I'm not sure what update to `main` you have in mind. Why remove `try_main`?

**Comment by dtolnay** [comment]
> How exactly do I test this, what does `cratesio-index` take as `<INDEX>`?

Ah sorry, the index is a clone of [github.com/rust-lang/crates.io-index](https://github.com/rust-lang/crates.io-index). So something like this:

```console
$ cd ~/path/to
$ git clone https://github.com/rust-lang/crates.io-index
$ cargo run --manifest-path ~/path/to/cargo-tally/index/Cargo.toml --release -- ~/path/to/crates.io-index
```

**Comment by dtolnay** [comment]
Nice! That graph at least has roughly the right shape, we just need data at every time dependencies change rather than only at releases of serde, as you identified.

> Am I correct in thinking that each crate needs to know every crate that depends on it at every point in time? This allows comparison without timestamp problems hopefully fixing my broken graph problem.

Yes, one way or another the precomputed data needs to contain enough information to efficiently figure out every crate that depends on every crate.

In the most extreme case, you can imagine emitting a complete snapshot of the entire transitive dependency graph at every timestamp that a crate is published. For example if the history consists of:

```
- at time T0, publish "aaa 0.1.0"
- at time T1, publish "bbb 0.1.0" that depends on "aaa ^0.1"
- at time T2, publish "ccc 0.1.0" that depends on "bbb ^0.1"
- at time T3, publish "aaa 0.2.0"
- at time T4, publish "bbb 0.1.1" that depends on "aaa ^0.2"
```

then the precomputed transitive reverse dependencies would be:

```
- T0: ∅
- T1: {aaa 0.1.0: [bbb 0.1.0]}
- T2: {aaa 0.1.0: [bbb 0.1.0, ccc 0.1.0], bbb 0.1.0: [ccc 0.1.0]}
- T3: {aaa 0.1.0: [bbb 0.1.0, ccc 0.1.0], bbb 0.1.0: [ccc 0.1.0]}
- T4: {aaa 0.2.0: [bbb 0.1.1, ccc 0.1.0], bbb 0.1.1: [ccc 0.1.0]}
```

From this precomputed data it's possible to efficiently answer cargo-tally queries like how many crates depend transitively on aaa, or transitively on aaa 0.1, including with --exclude flags taken into account.

In reality this naive representation would be quite large (maybe terabytes). If we record diffs instead, I think it will be reasonable. Notice that from T2 to T3 there are no changes in the global transitive dependency graph so T3 can actually be omitted entirely. And each crate release only affects a small part of the global graph, so only the parts that change need to go in the output.

With some gzip compression on top, I think the result will be on the order of a few dozen megabytes which is fine.

File: index/src/main.rs (hunk 2 above)
Comment (DevinR528) [reply index/src/main.rs]: I meant having `main` return the `Result` but after thinking more about, it try_main seems clearer.

Comment (dtolnay) [reply index/src/main.rs]: `main` should never return Result because that would make the error get Debug-printed, whereas Display is what I want here.

**Comment by DevinR528** [comment]
I'm running it as I type to make sure but I'm pretty sure i got it working. It seems to run about n^2 slower, and of course I lost power about 40% of they way through 🤦‍♂️ so I had to restart.

I will push once it finishes and I can graph something.

It will still need the serialized type to be rewritten but as long as the algorithm is correct that'll be trivial.

**Comment by DevinR528** [comment]
Does it make sense that its still running, I don't have the fastest computer (16g ram and a Amd Ryzen 3 2200g) there is about 200,00 events in the crates.io index I'm using? It is definitely quadratic in time complexity.
The algorithm is
```rust
for krate in crates
    process_event
    compute_counts
    add to table
    // from here down solves the timestamp/graph problem
    for (name, metas) in universe.crates
        for meta in metas
            re comput_counts
            find last computed and compare to re-computed if diff add to table
```

**Comment by dtolnay** [comment]
That sounds way slower than I would expect. Keep in mind that the vast majority of publishes only affect transitive dependencies of a tiny fraction of the crate graph so there shouldn't be an inner loop over all crates.

**Comment by DevinR528** [comment]
Ok I had a feeling and didn't want to waste a few days on a bad attempt!
Something like this should work then?
```rust 
for krate in crates
    // during the process_events update any crate that had transitive deps changed
    process_event
    compute_counts
    add to table
```
Is `redo` the crates that are changed in between version publishes, if so
this set is the only thing that would need to be updated correct?
```rust 
let mut redo = Set::default();
if let Some(prev) = self.crates.get(&event.name) {
    // events CrateName and index into Metadata in Universe.crates
    let prev_key = CrateKey::new(event.name, prev.len() as u32 - 1);
    for dep in &self.depends[&prev_key] {
        self.reverse_depends.get_mut(dep).unwrap().remove(&prev_key);
    }
    self.depends.remove(&prev_key);
    for (i, metadata) in prev.iter().enumerate() {
        // if event lies within 
        if compatible_req(&metadata.num).matches(&event.num) {
            let key = CrateKey::new(event.name, i as u32);
            for node in self.reverse_depends[&key].clone() {
                for dep in &self.depends[&node] {
                    self.reverse_depends.get_mut(dep).unwrap().remove(&node);
                }
                redo.insert(node);
            }
        }
    }
}
```

**Comment by DevinR528** [comment]
AHH dang sorry I got a to large .json file stuck in the git history and could not push up github so I tried a few things and learned the hard way that `git filter-branch ...` is not supposed to be used by someone unfamiliar with it.

I was trying to push up to ask what you thought of this
```rust
// the is inside process_event after removing found crates and pinning version nums
let mut to_update = Vec::new();

// index is which dependency we mean in terms of Metadata
let index = self.crates.entry(event.name).or_insert_with(Vec::new).len();
let key = CrateKey::new(event.name, index as u32);

// the updated crates dep's and trans dep's
let trans_res = self.resolve_and_add_to_graph(key, &metadata);
to_update.extend(trans_res.crates.keys());

self.reverse_depends.insert(key, Set::default());
self.crates.get_mut(&event.name).unwrap().push(metadata);

for outdated in redo.iter() {
    let metadata = self.crates[&outdated.name][outdated.index as usize].clone();
    warn!("re-resolving {} {}", outdated.name, metadata.num);

    // do I need to also recheck the redo crates??
    let re_res = self.resolve_and_add_to_graph(*outdated, &metadata);
    to_update.extend(re_res.crates.keys());
}
// returns only the updated crates to re count transitive deps at every
// time point they changed
to_update
```
so I'm only looping over crates that I have to re count because of update right? If this is the way to
do it it's been running for 3 days and is at 50%

I will open a new one with my changes at the current point

**Pull request closed without merging**